client/landing/stepper/hooks/use-site-copy.tsx

```diff
@@ -0,0 +1,79 @@
+import { WPCOM_FEATURES_COPY_SITE } from '@automattic/calypso-products';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useDispatch } from '@wordpress/data';
+import { useEffect, useMemo, useCallback } from 'react';
+import { useDispatch as useReduxDispatch, useSelector } from 'react-redux';
+import { useQuerySitePurchases } from 'calypso/components/data/query-site-purchases';
+import { ONBOARD_STORE } from 'calypso/landing/stepper/stores';
+import { clearSignupDestinationCookie } from 'calypso/signup/storageUtils';
+import { getCurrentUserId } from 'calypso/state/current-user/selectors';
+import {
+	hasLoadedSitePurchasesFromServer,
+	isFetchingSitePurchases,
+} from 'calypso/state/purchases/selectors';
+import siteHasFeature from 'calypso/state/selectors/site-has-feature';
+import { fetchSiteFeatures } from 'calypso/state/sites/features/actions';
+import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+
+function useSafeSiteHasFeature( siteId: number, feature: string ) {
+	const dispatch = useReduxDispatch();
+	useEffect( () => {
+		dispatch( fetchSiteFeatures( siteId ) );
+	}, [ dispatch, siteId ] );
+
+	return useSelector( ( state ) => {
+		return siteHasFeature( state, siteId, feature );
+	} );
+}
+
+export const useSiteCopy = ( site: SiteExcerptData ) => {
+	const userId = useSelector( ( state ) => getCurrentUserId( state ) );
+	const hasCopySiteFeature = useSafeSiteHasFeature( site.ID, WPCOM_FEATURES_COPY_SITE );
+	const plan = site?.plan;
+	const isSiteOwner = site.site_owner === userId;
+	useQuerySitePurchases( site.ID );
+	const isLoadingPurchase = useSelector(
+		( state ) => isFetchingSitePurchases( state ) || ! hasLoadedSitePurchasesFromServer( state )
+	);
+
+	const { setPlanCartItem } = useDispatch( ONBOARD_STORE );
+
+	const shouldShowSiteCopyItem = useCallback( () => {
+		return hasCopySiteFeature && isSiteOwner && plan && ! isLoadingPurchase;
+	}, [ hasCopySiteFeature, isSiteOwner, plan, isLoadingPurchase ] );
+
+	const startSiteCopy = useCallback(
+		( recordTracks: () => void ) => {
+			if ( ! plan ) {
+				return;
+			}
+			clearSignupDestinationCookie();
+			setPlanCartItem( { product_slug: plan.product_slug } );
+			recordTracks();
```

Comment:
If the function will always `recordTracks` , then I prefer to receive the eventName as string. Something like:

```suggestion
		( eventName: CopySiteEventName ) => {
			if ( ! plan ) {
				return;
			}
			clearSignupDestinationCookie();
			setPlanCartItem( { product_slug: plan.product_slug } );
			recordTracksEvent(eventName);
```

where `eventName` is a typed variable:

```
type CopySiteEventName = 'calypso_settings_copy_site_option_click' | 'another_event'
```

Reply:
I'm not sure about the function that is called in both cases (in SiteTools and in EllipsisMenu).
Site tools seem that it is using `recordTracksEvent` and ellipsis menu uses a `recordTrack` which is a prop, and I'm not familiar if it does something else except that records the event.

Reply:
Yes!,
`recordTrack` in `sites-ellipsis-menu.tsx` uses `recordTracksEvent` under the hood.

https://github.com/Automattic/wp-calypso/blob/76bedd060f565be562d8428ec758e5eb0928f2fe/client/sites-dashboard/components/sites-ellipsis-menu.tsx#L263-L268

Reply:
The case I would like to avoid is that `recordTracks` callback function starts being used for more things than just "tracking events".

Reply:
Ok then! I'll refactor it to just pass the event instead of passing a callback! Thanks for clarifying that!

Reply:
> type CopySiteEventName = 'calypso_settings_copy_site_option_click' | 'another_event'

I left it as as string so someone would not have to update it every time this is used. It could probably become a monstrous type ( not many chances though).
Also, leaving it as a string, matches what we already have in ellipsis menu.